onnxruntime/python/tools/transformers/optimizer_utils.py

```diff
@@ -0,0 +1,34 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation.  All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+from fusion_utils import NumpyHelper
+from onnx import ModelProto
+from onnx.external_data_helper import set_external_data
+
+from onnxruntime import OrtValue
+
+
+def extract_external_data_from_model(model: ModelProto):
+    """
+    Extract external data from model and return the external data as a list of tuples (name, value).
+
+    Args:
+        model (ModelProto): the model proto to extract external data from.
+    Returns:
+        (external_names, external_values): a tuple of two lists of external data names and values.
+    """
+    external_data = []
+    for tensor in model.graph.initializer:
+        name = tensor.name
+
+        if tensor.HasField("raw_data"):
```

Comment:
This function has not handled sub-graph. Shall we check and throw exception if there is subgraph?

Reply:
Does initializer have sub-graph?

Reply:
An attribute of a node like If/Loop/BeamSearch can be ModelProto. That ModelProto can still have initializers.